File: topcoder/tccc04/BadNeighbors.py

memo = dict()


def f(donations):
    if(tuple(donations) in memo):
        return memo[tuple(donations)]
    if(len(donations) == 0):
        return 0
    if(len(donations) < 3):
        return max(donations)
    if(len(donations) == 3):
        return max(donations[0] + donations[2], donations[1])
    memo[tuple(donations)] = max(donations[0] + f(donations[2:]),
                                 f(donations[1:]))
    return memo[tuple(donations)]


class BadNeighbors:
    def maxDonations(self, donations):
        return max(f(donations[1:-1]), donations[0] + f(donations[2:-1]),
                   donations[-1] + f(donations[1:-2]))

File: topcoder/tccc04/BadNeighbors.cpp
#include <bits/stdc++.h>

using namespace std;

int dp[50][2];

class BadNeighbors {
public:
	int maxDonations(vector <int> donations) {
		int n = donations.size();
		dp[0][0] = 0;
		dp[0][1] = donations[0];
		dp[1][0] = donations[1];
		dp[1][1] = max(donations[0], donations[1]);
		for(int i = 2; i < n; i++){
			int &money = donations[i];
			dp[i][0] = max(money + dp[i-2][0], dp[i-1][0]);
			dp[i][1] = max(money + dp[i-2][1], dp[i-1][1]);
		}
		return max(dp[n-1][0], dp[n-2][1]);
	}
};
